account_financial_report_webkit/models/account_static_balance.py

```diff
@@ -53,119 +61,179 @@ def unlink(self):
         closed periods and accounts if their combination is missing. """
-        periods, accounts = self.get_missing_periods_and_accounts()
+        periods, accounts, journals = \
+            self.get_missing_periods_accounts_and_journals()
         if not periods:
             logger.debug("No missing entries found. Static account balance "
                          "data is up to date.")
             return
-        self.calculate_static_balance(periods=periods, accounts=accounts)
+        self.calculate_static_balance(
+            periods=periods, accounts=accounts, journals=journals)
 
     @api.multi
-    def get_missing_periods_and_accounts(self, periods=None):
+    def get_missing_periods_accounts_and_journals(self, periods=None):
         # Split into unique periods and accounts
-        per_ids = acc_ids = []
+        per_ids = []
+        acc_ids = []
+        journal_ids = []
         missing_combinations = self.get_entries_to_calculate(periods=periods)
         if missing_combinations:
-            per_ids, acc_ids = map(list, map(set, zip(*missing_combinations)))
+            per_ids, acc_ids, journal_ids = map(
+                list, map(set, zip(*missing_combinations)))
         missing_periods = self.env['account.period'].browse(per_ids)
         missing_accounts = self.env['account.account'].browse(acc_ids)
-        return missing_periods, missing_accounts
+        missing_journals = self.env['account.journal'].browse(journal_ids)
+        return missing_periods, missing_accounts, missing_journals
 
     @api.model
-    def calculate_static_balance(self, periods, accounts=None):
+    def calculate_static_balance(self, periods, accounts=None, journals=None):
         """ Calculate the balance for a specified set of periods
         and insert them into the table."""
         self = self.suspend_security()
-        periods = periods.filtered(
-            lambda p: p.state == 'done' and not p.special)
-        for row in self.calculate_balance(periods, accounts=accounts):
-            try:
-                with self.env.cr.savepoint():
-                    self.create(row)
-            except psycopg2.IntegrityError:
-                # A balance already exists for this period and account
-                pass
+
+        if not journals:
+            # When journals are set, we can assume that this is a close for
+            # a journal period. Otherwise we need to filter the periods to
+            # check if they are in a done state
+            periods = periods.filtered(
+                lambda p: p.state == 'done' and not p.special)
+
+        for period in periods:
+            for row in self.calculate_balance(
+                    period, accounts=accounts, journals=journals):
+                try:
+                    with self.env.cr.savepoint():
+                        self.create(row)
+                except psycopg2.IntegrityError as e:
+                    # A balance already exists for this period and account
```

Comment:
Wondering if this is a design flaw in the original version. It seems to me now that an existing balance might be outdated.

Reply:
I dont think so, because when you reopen a period, the static balances are unlinked. But in this version i added an unlink filters for existing accounts: https://github.com/EssentNovaTeam/account-financial-reporting/pull/6/files/b7c2bd583ebbe837b1e90d75d62d61198d78121b#diff-c4749f6b2ebab521a3de0b3add26b25c49bc78e8c8a4271f430e91f2cf6b4f54R193

Reply:
OK, I've been milling on this all night and the scenario that I think could be a problem is when a journal period is reopened before the previous job runs. Reopening would delete the existing balances, but they have not been generated yet. When they are generated, they are outdated. When the journal period is closed again, recalculation of its balances are triggered but not stored due to the IntegrityError.
Mitigation we can apply: if the recomputation is triggered specifically with a set of journals, unlink existing balances, and when storing balances, skip the balances for journal periods that are not in 'done' state. Or maybe you can think of other, or better approaches?

Reply:
One wild approach that comes to mind is to add a dummy flag to the journal period model that is set to value 'a' when reopened, value 'b' when its closed, and value 'c' when the balances are stored. If the journal period status was changed while balances are being generated it would raise a concurrency error in the longest running of the competing transactions (creation of the balances) which can be caught there to throw a RetryableJobError. Maybe combine with a commit after storing the balances for a specific period.

Reply:
Implemented an integrity validation method: https://github.com/EssentNovaTeam/account-financial-reporting/pull/6/commits/e4a3cdb00f3c2e7195db389d1a8e616f33e6dd4c#diff-c4749f6b2ebab521a3de0b3add26b25c49bc78e8c8a4271f430e91f2cf6b4f54R295-R321 (also tested in test05)

Reply:
I still think you should unlink any existing balance at this point.

Reply:
Any, or only invalid balances?

Reply:
Any balance that will trigger the IntegrityError because it covers the same period, journal and account.

Reply:
Check, on it!

Reply:
Added the unlink: https://github.com/EssentNovaTeam/account-financial-reporting/pull/6/commits/700b90476b46e3a6f2ce7234c493130bbd0bca42#diff-c4749f6b2ebab521a3de0b3add26b25c49bc78e8c8a4271f430e91f2cf6b4f54R131-R135